File: source/bindings/csharp/speech_synthesis_word_boundary_event_args.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
using System;
using System.Globalization;
using Microsoft.CognitiveServices.Speech.Internal;
using static Microsoft.CognitiveServices.Speech.Internal.SpxExceptionThrower;

namespace Microsoft.CognitiveServices.Speech
{
    /// <summary>
    /// Define payload of speech synthesis events.
    /// Added in version 1.7.0
    /// </summary>
    public class SpeechSynthesisWordBoundaryEventArgs : System.EventArgs, IDisposable
    {
        internal SpeechSynthesisWordBoundaryEventArgs(IntPtr eventHandlePtr)
        {
            eventHandle = new InteropSafeHandle(eventHandlePtr, Internal.Synthesizer.synthesizer_event_handle_release);

            ThrowIfFail(Internal.Synthesizer.synthesizer_word_boundary_event_get_values(eventHandlePtr, out audioOffset, out textOffset, out wordLength));
        }

        /// <summary>
        /// Specifies current word's binary offset in output audio, by ticks (100ns).
        /// </summary>
        public UInt64 AudioOffset
        {
            get
            {
                return audioOffset;
            }
        }

        /// <summary>
        /// Specifies current word's text offset in input text, by characters.
        /// </summary>
        public UInt32 TextOffset
        {
            get
            {
                return textOffset;
            }
        }

        /// <summary>
        /// Specifies current word's length, by characters.
        /// </summary>
        public UInt32 WordLength
        {
            get
            {
                return wordLength;
            }
        }

        private InteropSafeHandle eventHandle;

        private UInt64 audioOffset;

        private UInt32 textOffset;

        private UInt32 wordLength;

        internal volatile bool disposed = false;

        /// <summary>
        /// Dispose of associated resources.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// This method performs cleanup of resources.
        /// The Boolean parameter <paramref name="disposing"/> indicates whether the method is called from <see cref="IDisposable.Dispose"/> (if <paramref name="disposing"/> is true) or from the finalizer (if <paramref name="disposing"/> is false).
        /// Derived classes should override this method to dispose resource if needed.
        /// </summary>
        /// <param name="disposing">Flag to request disposal.</param>
        /// <returns></returns>
        protected virtual void Dispose(bool disposing)
        {
            if (disposed)
            {
                return;
            }

            if (disposing)
            {
                // free managed resources
                eventHandle?.Dispose();
            }
            // free native resources if there are any.
            disposed = true;
        }
    }
}

File: source/bindings/java/com/microsoft/cognitiveservices/speech/SpeechSynthesisWordBoundaryEventArgs.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
package com.microsoft.cognitiveservices.speech;

import java.io.Closeable;

import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;
import com.microsoft.cognitiveservices.speech.util.SafeHandleType;
import com.microsoft.cognitiveservices.speech.util.IntRef;

/**
 * Defines contents of speech synthesis word boundary event.
 * Added in version 1.7.0
 */
public class SpeechSynthesisWordBoundaryEventArgs {

    /*! \cond INTERNAL */

    SpeechSynthesisWordBoundaryEventArgs(long eventArgs) {
        Contracts.throwIfNull(eventArgs, "eventArgs");
        
        SafeHandle eventHandle = new SafeHandle(eventArgs, SafeHandleType.SynthesisEvent);
        IntRef audioOffsetRef = new IntRef(0);
        IntRef textOffsetRef = new IntRef(0);
        IntRef wordLengthRef = new IntRef(0);
        Contracts.throwIfFail(getWordBoundaryEventValues(eventHandle, audioOffsetRef, textOffsetRef, wordLengthRef));
        this.audioOffset = audioOffsetRef.getValue();
        this.textOffset = textOffsetRef.getValue();
        this.wordLength = wordLengthRef.getValue();
        eventHandle.close();
    }

    /*! \endcond */
    
    /**
     * Specifies current word's binary offset in output audio, by ticks (100ns).
     * @return Current word's binary offset in output audio, by ticks (100ns).
     */
    public long getAudioOffset() {
        return this.audioOffset;
    }

    /**
     * Specifies current word's text offset in input text, by characters.
     * @return Current word's text offset in input text, by characters.
     */
    public long getTextOffset() {
        return this.textOffset;
    }

    /**
     * Specifies current word's length, by characters.
     * @return Current word's length, by characters.
     */
    public long getWordLength() {
        return this.wordLength;
    }

    private final native long getWordBoundaryEventValues(SafeHandle eventHandle, IntRef audioOffset, IntRef textOffset, IntRef wordLength);

    private long audioOffset;
    private long textOffset;
    private long wordLength;

}
